packages/java-parser/src/index.js

```diff
@@ -43,10 +46,71 @@ function parse(inputText, entryPoint = "compilationUnit") {
         error.context.ruleStack.join("\n\t->")
     );
   }
-
   return cst;
 }
 
+function attachComments(tokenArray) {
+  const tokens = tokenArray;
+  const newTokens = [];
+  const headcomments = [];
+  let i = 0;
+
+  // pushing comments at the beginning of the file to the leadingComments of the first non Comment token
+  while (tokenMatcher(tokens[i].tokenType, t.Comment) && i < tokens.length) {
```

Comment:
would not it be simpler to collect the comments in a separate **group** and 
run over two indexes, one for the comments and one for the regular tokens?

https://sap.github.io/chevrotain/docs/features/token_grouping.html

There seems like a great deal of edge cases and conditionals in this logic, this can easily lead to bugs.

Reply:
I guess we could do that way.

Reply:
I pushed a new version of it, using the group comment.

Reply:
i will have a look tomorrow probably